compiler/rustc_ast/src/lib.rs

```diff
@@ -15,6 +15,7 @@
 #![feature(if_let_guard)]
 #![feature(let_chains)]
 #![feature(min_specialization)]
+#![feature(spec_option_partial_eq)]
```

Comment:
You don't need to enable this feature in a crate to get the benefits of the specialization -- that's not how specialization works, right?

Reply:
The `newtype` macro expansion contains an specialised implementation of the trait, so as I understand it, the crates using the macro need to enable the feature to be able to access the trait.

Reply:
`impl`s are universal -- stability attributes don't really matter on them, so `#![feature]` is not needed here. For proof, I checked out your branch, compiled down this code (without the feature enabled):

```rust
use std::num::NonZeroUsize;

pub fn match1(l: Option<NonZeroUsize>, r: Option<NonZeroUsize>) -> bool {
    l == r
}
```

(`rustc +rust4 ~/test.rs --emit=llvm-ir --crate-type=lib -O -o test.opt.ll`)

And got:
```llvm
define noundef zeroext i1 @_ZN4test6match117h4797d563345a25cdE(i64 %0, i64 %1) unnamed_addr #0 {
start:
  %2 = icmp eq i64 %0, %1
  ret i1 %2
}
```

Reply:
Cf. nightly (`rustc +nightly ~/test.rs --emit=llvm-ir --crate-type=lib -O -o test.orig.ll`)

```llvm
define noundef zeroext i1 @_ZN4test6match117h0f92982f5e9290cbE(i64 %0, i64 %1) unnamed_addr #0 {
start:
  %2 = icmp ne i64 %0, 0
  %3 = icmp ne i64 %1, 0
  %4 = xor i1 %2, %3
  br i1 %4, label %"_ZN70_$LT$core..option..Option$LT$T$GT$$u20$as$u20$core..cmp..PartialEq$GT$2eq17hd2ff3c6638277712E.exit", label %bb2.i

bb2.i:                                            ; preds = %start
  %5 = select i1 %2, i1 %3, i1 false
  %6 = icmp eq i64 %0, %1
  %not..i = xor i1 %5, true
  %spec.select.i = select i1 %not..i, i1 true, i1 %6
  br label %"_ZN70_$LT$core..option..Option$LT$T$GT$$u20$as$u20$core..cmp..PartialEq$GT$2eq17hd2ff3c6638277712E.exit"

"_ZN70_$LT$core..option..Option$LT$T$GT$$u20$as$u20$core..cmp..PartialEq$GT$2eq17hd2ff3c6638277712E.exit": ; preds = %start, %bb2.i
  %.0.i = phi i1 [ false, %start ], [ %spec.select.i, %bb2.i ]
  ret i1 %.0.i
}
```

Reply:
WAIT, I think I understand what you mean. :facepalm: Marked my other comments as off-topic. I didn't realize you meant `newtype_index!`.

Reply:
Though I guess it's a shame that you can't gate the usage of `SpecOptionPartialEq` silently in the derive macro itself like `#[derive(Eq)]` does with `StructuralEq` :/

Reply:
Does macro hygiene not cover features?

Reply:
Would adding `spec_option_partial_eq` here help?
https://github.com/rust-lang/rust/blob/2afca78a0b03db144c5d8b9f8868feebfe096309/compiler/rustc_macros/src/lib.rs#L50